### ISSUE TITLE
TX indexing keys should be managed by app and not tendermint

### ISSUE THREAD

**Issue by ebuchman** [tendermint/tendermint#525]
Right now we hash the whole transaction for the indexing, which could lead to malleability issues for non-deterministic signature schemes.

Perhaps this is a great use-case for key-value pairs in the DeliverTx response

**Comment by srmo** [comment]
The most important part in this is that we need the hash that can be used to use the chain's query interface. At this level, it doesn't matter if hashes are generated by core or by the app.
It starts to matter when we think about core chaning it's hashing implementation: if the app persisted a list of hashes, how do we migrate it to a new scheme? Does it even make sense to think about changing the hashing implementation?

Having the hash generation in the app would give full control to the implementers. Then again, is there a need to make sure implementers can give guarantees on the quality of the hashing function?

Either way, having it in one place only is the important part and giving the app easy access to the hash too. What good is it to know that you can query by hash but the app never sees the hash or has to keep up with duplicating the hash implementation.

**Comment by srmo** [comment]
using the key value approach could also offer other benefits. What about also returning a key-value pair for the index/position in the block? IIRC, at the moment if a tx was in the 2nd deliverTx call after BeginBlock we know it will be at position 2 in the block.
But will this always hold true? What about making deliverTx work in parallel so we can handle multiple deliverTx streams at the same time and therefor increase throughput on the app side - given that this puts additional burden on the app implementers - multiThreading óle!

**Comment by melekes** [comment]
> or make it possible to choose between sequential/parallel deliverTx

might be worth looking into this, but that's out of the scope of this ticket. note: some people rely on order so if we're gonna do it, it should be a config option for sure.

> What good is it to know that you can query by hash but the app never sees the hash or has to keep up with duplicating the hash implementation.

:+1: 

What is the use case for an app duplicating the hashing impl? because it wants to store some additional details?

**Comment by srmo** [comment]
> But what is the use case for an app duplicating the hashing impl? because it wants to store some additional details?
Let me link to https://github.com/tendermint/tendermint/issues/423 where the basic use case is described (if not, let me know).
So the idea is that we need to know the TX hash in deliverTx (maybe even in CheckTX) so we can have our own indizes for quick lookups. Think like having a "CREATE INDEX ..." that creates a index local to the app with the blockChain as the DB, so our lookups can use different local indices to retrieve the required hashes that can then be used to query the chain for the actual TXs.

Does that make sense?

**Comment by srmo** [comment]
To clarify "lookups"
we need to write queries like `give transaction of type X where attribute1 < n and attribute2 >= m`
Now, we could have a denormalized persistence store for transactions, but this is doubling the persistence, rather have multiple indices will return hashes for such queries that you then use to query the blockchain with. There you need the tx hash as it is in the chain, as early as possible.

**Comment by ebuchman** [comment]
Definitely see the need for more advanced indexing in the app eg. by attribute. We could add this to Tendermint using the tags and a more advanced indexer but I'm afraid it might be a burden of too many feature requests since every app will probably want to do things somewhat differently. Or is this something we could reasonably handle all Tendermint side? If Tendermint could provide the full range of indexing, the app wouldnt need the tx hash, right?

**Comment by niniwzw** [comment]
the indexing like ethereum transaction reciept and event， maybe indexer can do like ethereum

**Comment by srmo** [comment]
I don't see a way how tendermint will be able to implement it in a way that any app will be happy with it. From my POV you would need to implement a full query engine that allows you to create arbitrarily complex queries and then you also need the required indeices for such queries. In the end, this looks like you'd need to implement a DBMS and that's certainly out of scope.
While I'd be happy to have such a thing in core, I don't think you can provide the required functionality as easy as an app would. 
So it would be easier/faster to provide the app with the tx hash as soon as possible, probably already at checkTx time?

But yes, if core (or a module that can be activated) provides *full* indexing and query capabilities everything should be fine. Please take my comment with a grain of salt that I didn't think through how you could implement such a thing. Maybe it's easy after all and the query interface works like a NoSQL engine, as we "just" need queries on key/value pairs.

**Comment by melekes** [comment]
Indexing proposal https://gist.github.com/melekes/131291e5144e31d34ac3ed3ab7ff2aca

**Comment by ebuchman** [comment]
Think this is closed now with 0.13 and the new tx query interface. We will be looking to extend this out with an optional SQL database on tendermint side soon too. Please open a new issue if 0.13 doesn't meet your needs

**Comment by ebuchman** [comment]
Forgot to close it!
